Problem compiling darktable from git since commit ccdb5042a13b6ad88923716622fbc546bbeb38e2

Please bear with me if the info is missing something, just ask. I'm neither a programmer nor a git expert ...

On Slackware64-current I experience a build error when compiling darktable from the git sources. I hunted it down by checking out the different commits and found commit ccdb5042a13b6ad88923716622fbc546bbeb38e2 to be the culprit for me.

Error message I get when compiling (see also the attached longer make.log):

[ 18%] Building C object src/CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o
cd /tmp/SBo/darktable/build/src && /usr/bin/clang -DGDK_DISABLE_DEPRECATED -DGDK_VERSION_MIN_REQUIRED=GDK_VERSION_3_22 -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_MIN_REQUIRED -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_40 -DGTK_DISABLE_DEPRECATED -DGTK_DISABLE_SINGLE_INCLUDES -DHAVE_BUILTIN_CPU_SUPPORTS -DHAVE_CONFIG_H -DHAVE_GPHOTO2 -DHAVE_GPHOTO_25_OR_NEWER -DHAVE_HTTP_SERVER -DHAVE_ISO_CODES -DHAVE_KWALLET -DHAVE_LENSFUN -DHAVE_LIBSECRET -DHAVE_OPENCL -DHAVE_OPENEXR -DHAVE_OPENJPEG -DHAVE_PRINT -DUSE_LUA -D_XOPEN_SOURCE=700 -D__GDK_KEYSYMS_COMPAT_H__ -Dlib_darktable_EXPORTS -I/tmp/SBo/darktable/build/src -I/tmp/SBo/darktable/src -I/tmp/SBo/darktable/src/external/lua/src -I/tmp/SBo/darktable/src/external/LuaAutoC -I/tmp/SBo/darktable/build/src/external/rawspeed/src -I/tmp/SBo/darktable/src/external/rawspeed/src/librawspeed -isystem /tmp/SBo/darktable/src/external -isystem /tmp/SBo/darktable/src/external/OpenCL -isystem /tmp/SBo/darktable/src/external/rawspeed/src/external -isystem /usr/include/glib-2.0 -isystem /usr/lib64/glib-2.0/include -isystem /usr/include/gtk-3.0 -isystem /usr/include/at-spi2-atk/2.0 -isystem /usr/include/at-spi-2.0 -isystem /usr/include/dbus-1.0 -isystem /usr/lib64/dbus-1.0/include -isystem /usr/include/gio-unix-2.0 -isystem /usr/include/cairo -isystem /usr/include/pango-1.0 -isystem /usr/include/fribidi -isystem /usr/include/atk-1.0 -isystem /usr/include/pixman-1 -isystem /usr/include/uuid -isystem /usr/include/freetype2 -isystem /usr/include/harfbuzz -isystem /usr/include/libpng16 -isystem /usr/include/gdk-pixbuf-2.0 -isystem /usr/include/libmount -isystem /usr/include/blkid -isystem /usr/include/libxml2 -isystem /usr/include/libsoup-2.4 -isystem /usr/include/OpenEXR -isystem /usr/include/lensfun -isystem /usr/include/librsvg-2.0 -isystem /usr/include/json-glib-1.0 -isystem /usr/include/openjpeg-2.3 -isystem /usr/include/libsecret-1  -O2 -fPIC -fopenmp-version=45 -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -Wthread-safety -Wno-unknown-pragmas -Wno-error=varargs -Wno-error=address-of-packed-member -std=c99 -fopenmp=libomp -mtune=generic -msse2 -g -O3 -DNDEBUG -O3 -fPIC   -Werror -Wno-error=pass-failed -o CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o   -c /tmp/SBo/darktable/src/bauhaus/bauhaus.c
In file included from /tmp/SBo/darktable/src/bauhaus/bauhaus.c:20:
In file included from /tmp/SBo/darktable/src/bauhaus/bauhaus.h:23:
In file included from /tmp/SBo/darktable/src/control/control.h:34:
In file included from /tmp/SBo/darktable/src/libs/lib.h:23:
/tmp/SBo/darktable/src/views/view.h:270:3: error: unknown type name 'dt_history_copy_item_t'
  dt_history_copy_item_t copy_paste;
  ^
1 error generated.
make[2]: *** [src/CMakeFiles/lib_darktable.dir/build.make:82: src/CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o] Error 1
make[2]: Leaving directory '/tmp/SBo/darktable/build'
make[1]: *** [CMakeFiles/Makefile2:2284: src/CMakeFiles/lib_darktable.dir/all] Error 2
make[1]: Leaving directory '/tmp/SBo/darktable/build'
make: *** [Makefile:152: all] Error 2



cmake is at version 3.16.5, gcc at 9.3.0

When I checkout the commit ccdb5042a13b6ad88923716622fbc546bbeb38e2 and the commit before (d39ba3b8f71badc9bb025e137f6eaab28d6b0183) and compare the file src/views/view.h from both checkouts I'm seeing far more changes then shown in the commit message. Especially the mentioned 'dt_history_copy_item_t' is touched allthough it is not mentioned in the commit log.

Is this an error on my side, a git error or something else? Any clue how I can make it compile successfully?

[cmake.log](https://github.com/darktable-org/darktable/files/4353397/cmake.log)
[make.log](https://github.com/darktable-org/darktable/files/4353398/make.log)

Well I had tried with commit 3be2a4bfba5614365af93f365688e575b8fd33cc and worked my way backwards, but it didn't compile.

Adding the struct keyword as suggested by AlicVB changed the error message though, but the compilation ended unsuccessfully:

[ 18%] Building C object src/CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o
cd /tmp/SBo/darktable/build/src && /usr/bin/clang -DGDK_DISABLE_DEPRECATED -DGDK_VERSION_MIN_REQUIRED=GDK_VERSION_3_22 -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_MIN_REQUIRED -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_40 -DGTK_DISABLE_DEPRECATED -DGTK_DISABLE_SINGLE_INCLUDES -DHAVE_BUILTIN_CPU_SUPPORTS -DHAVE_CONFIG_H -DHAVE_GAME -DHAVE_GPHOTO2 -DHAVE_GPHOTO_25_OR_NEWER -DHAVE_HTTP_SERVER -DHAVE_ISO_CODES -DHAVE_KWALLET -DHAVE_LENSFUN -DHAVE_LIBSECRET -DHAVE_OPENCL -DHAVE_OPENEXR -DHAVE_OPENJPEG -DHAVE_PRINT -DHAVE_SQLITE_324_OR_NEWER -DUSE_LUA -D_XOPEN_SOURCE=700 -D__GDK_KEYSYMS_COMPAT_H__ -Dlib_darktable_EXPORTS -I/tmp/SBo/darktable/build/src -I/tmp/SBo/darktable/src -I/tmp/SBo/darktable/src/external/lua/src -I/tmp/SBo/darktable/src/external/LuaAutoC -I/tmp/SBo/darktable/build/src/external/rawspeed/src -I/tmp/SBo/darktable/src/external/rawspeed/src/librawspeed -isystem /tmp/SBo/darktable/src/external -isystem /tmp/SBo/darktable/src/external/OpenCL -isystem /tmp/SBo/darktable/src/external/rawspeed/src/external -isystem /usr/include/glib-2.0 -isystem /usr/lib64/glib-2.0/include -isystem /usr/include/gtk-3.0 -isystem /usr/include/at-spi2-atk/2.0 -isystem /usr/include/at-spi-2.0 -isystem /usr/include/dbus-1.0 -isystem /usr/lib64/dbus-1.0/include -isystem /usr/include/gio-unix-2.0 -isystem /usr/include/cairo -isystem /usr/include/pango-1.0 -isystem /usr/include/fribidi -isystem /usr/include/atk-1.0 -isystem /usr/include/pixman-1 -isystem /usr/include/uuid -isystem /usr/include/freetype2 -isystem /usr/include/harfbuzz -isystem /usr/include/libpng16 -isystem /usr/include/gdk-pixbuf-2.0 -isystem /usr/include/libmount -isystem /usr/include/blkid -isystem /usr/include/libxml2 -isystem /usr/include/libsoup-2.4 -isystem /usr/include/OpenEXR -isystem /usr/include/lensfun -isystem /usr/include/librsvg-2.0 -isystem /usr/include/json-glib-1.0 -isystem /usr/include/openjpeg-2.3 -isystem /usr/include/libsecret-1  -O2 -fPIC -fopenmp-version=45 -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -Wthread-safety -Wno-unknown-pragmas -Wno-error=varargs -Wno-error=address-of-packed-member -std=c99 -fopenmp=libomp -mtune=generic -msse2 -g -O3 -DNDEBUG -O3 -fPIC   -Werror -Wno-error=pass-failed -o CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o   -c /tmp/SBo/darktable/src/bauhaus/bauhaus.c
In file included from /tmp/SBo/darktable/src/bauhaus/bauhaus.c:19:
In file included from /tmp/SBo/darktable/src/bauhaus/bauhaus.h:22:
In file included from /tmp/SBo/darktable/src/control/control.h:33:
In file included from /tmp/SBo/darktable/src/libs/lib.h:22:
/tmp/SBo/darktable/src/views/view.h:268:33: error: field has incomplete type 'struct dt_history_copy_item_t'
  struct dt_history_copy_item_t copy_paste;
                                ^
/tmp/SBo/darktable/src/views/view.h:268:10: note: forward declaration of 'struct dt_history_copy_item_t'
  struct dt_history_copy_item_t copy_paste;
         ^
1 error generated.
make[2]: *** [src/CMakeFiles/lib_darktable.dir/build.make:82: src/CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o] Error 1
make[2]: Leaving directory '/tmp/SBo/darktable/build'
make[1]: *** [CMakeFiles/Makefile2:2317: src/CMakeFiles/lib_darktable.dir/all] Error 2
make[1]: Leaving directory '/tmp/SBo/darktable/build'
make: *** [Makefile:152: all] Error 2

Yes, it's the same with commit 64d5ee2d1b282aeee5adc4764adc20b847038029 and without the struct keyword. To be precise it's the same error about unknown struct. I upgraded to git 2.25.2 and did a completely new "git clone" just to be sure. The error:

[ 18%] Building C object src/CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o
cd /tmp/SBo/darktable/build/src && /usr/bin/clang -DGDK_DISABLE_DEPRECATED -DGDK_VERSION_MIN_REQUIRED=GDK_VERSION_3_22 -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_MIN_REQUIRED -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_40 -DGTK_DISABLE_DEPRECATED -DGTK_DISABLE_SINGLE_INCLUDES -DHAVE_BUILTIN_CPU_SUPPORTS -DHAVE_CONFIG_H -DHAVE_GAME -DHAVE_GPHOTO2 -DHAVE_GPHOTO_25_OR_NEWER -DHAVE_HTTP_SERVER -DHAVE_ISO_CODES -DHAVE_KWALLET -DHAVE_LENSFUN -DHAVE_LIBSECRET -DHAVE_OPENCL -DHAVE_OPENEXR -DHAVE_OPENJPEG -DHAVE_PRINT -DHAVE_SQLITE_324_OR_NEWER -DUSE_LUA -D_XOPEN_SOURCE=700 -D__GDK_KEYSYMS_COMPAT_H__ -Dlib_darktable_EXPORTS -I/tmp/SBo/darktable/build/src -I/tmp/SBo/darktable/src -I/tmp/SBo/darktable/src/external/lua/src -I/tmp/SBo/darktable/src/external/LuaAutoC -I/tmp/SBo/darktable/build/src/external/rawspeed/src -I/tmp/SBo/darktable/src/external/rawspeed/src/librawspeed -isystem /tmp/SBo/darktable/src/external -isystem /tmp/SBo/darktable/src/external/OpenCL -isystem /tmp/SBo/darktable/src/external/rawspeed/src/external -isystem /usr/include/glib-2.0 -isystem /usr/lib64/glib-2.0/include -isystem /usr/include/gtk-3.0 -isystem /usr/include/at-spi2-atk/2.0 -isystem /usr/include/at-spi-2.0 -isystem /usr/include/dbus-1.0 -isystem /usr/lib64/dbus-1.0/include -isystem /usr/include/gio-unix-2.0 -isystem /usr/include/cairo -isystem /usr/include/pango-1.0 -isystem /usr/include/fribidi -isystem /usr/include/atk-1.0 -isystem /usr/include/pixman-1 -isystem /usr/include/uuid -isystem /usr/include/freetype2 -isystem /usr/include/harfbuzz -isystem /usr/include/libpng16 -isystem /usr/include/gdk-pixbuf-2.0 -isystem /usr/include/libmount -isystem /usr/include/blkid -isystem /usr/include/libxml2 -isystem /usr/include/libsoup-2.4 -isystem /usr/include/OpenEXR -isystem /usr/include/lensfun -isystem /usr/include/librsvg-2.0 -isystem /usr/include/json-glib-1.0 -isystem /usr/include/openjpeg-2.3 -isystem /usr/include/libsecret-1  -O2 -fPIC -fopenmp-version=45 -Wall -Wformat -Wformat-security -Wshadow -Wtype-limits -Wvla -Wthread-safety -Wno-unknown-pragmas -Wno-error=varargs -Wno-error=address-of-packed-member -std=c99 -fopenmp=libomp -mtune=generic -msse2 -g -O3 -DNDEBUG -O3 -fPIC   -Werror -Wno-error=pass-failed -o CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o   -c /tmp/SBo/darktable/src/bauhaus/bauhaus.c
In file included from /tmp/SBo/darktable/src/bauhaus/bauhaus.c:19:
In file included from /tmp/SBo/darktable/src/bauhaus/bauhaus.h:22:
In file included from /tmp/SBo/darktable/src/control/control.h:33:
In file included from /tmp/SBo/darktable/src/libs/lib.h:22:
/tmp/SBo/darktable/src/views/view.h:268:3: error: unknown type name 'dt_history_copy_item_t'
  dt_history_copy_item_t copy_paste;
  ^
1 error generated.
make[2]: *** [src/CMakeFiles/lib_darktable.dir/build.make:82: src/CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o] Error 1
make[2]: Leaving directory '/tmp/SBo/darktable/build'
make[1]: *** [CMakeFiles/Makefile2:2317: src/CMakeFiles/lib_darktable.dir/all] Error 2
make[1]: Leaving directory '/tmp/SBo/darktable/build'
make: *** [Makefile:152: all] Error 2

Without "submodule init" rawspeed module is not found. I do not "git pull", but cloned the git tree and created a tar.gz archive from it, that I'm extracting for every compilation.

The output is looking slightly different, less verbose, but I guess it is in the end the same error. This is without any optimization, SLKFLAGS or other fancy stuff at all:

[ 23%] Built target rawspeed
Scanning dependencies of target rawspeed_get_number_of_processor_cores
[ 23%] Building CXX object src/external/rawspeed/src/librawspeed/common/CMakeFiles/rawspeed_get_number_of_processor_cores.dir/GetNumberOfProcessorCores.cpp.o
[ 23%] Linking CXX static library librawspeed_get_number_of_processor_cores.a
[ 23%] Built target rawspeed_get_number_of_processor_cores
Scanning dependencies of target darktable-rs-identify
[ 25%] Building CXX object src/external/rawspeed/src/utilities/identify/CMakeFiles/darktable-rs-identify.dir/rawspeed-identify.cpp.o
Scanning dependencies of target lib_darktable
[ 25%] Building C object src/CMakeFiles/lib_darktable.dir/version_gen.c.o
[ 25%] Building C object src/CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o
[ 25%] Building C object src/CMakeFiles/lib_darktable.dir/common/bilateral.c.o
[ 25%] Building C object src/CMakeFiles/lib_darktable.dir/common/bilateralcl.c.o
[ 25%] Linking CXX executable darktable-rs-identify
In file included from /tmp/SBo/darktable/src/libs/lib.h:22,
                 from /tmp/SBo/darktable/src/control/control.h:33,
                 from /tmp/SBo/darktable/src/bauhaus/bauhaus.h:22,
                 from /tmp/SBo/darktable/src/bauhaus/bauhaus.c:19:
/tmp/SBo/darktable/src/views/view.h:268:3: error: unknown type name ‘dt_history_copy_item_t’
  268 |   dt_history_copy_item_t copy_paste;
      |   ^~~~~~~~~~~~~~~~~~~~~~
gmake[2]: *** [src/CMakeFiles/lib_darktable.dir/build.make:82: src/CMakeFiles/lib_darktable.dir/bauhaus/bauhaus.c.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [CMakeFiles/Makefile2:2273: src/CMakeFiles/lib_darktable.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
[ 25%] Built target darktable-rs-identify
gmake: *** [Makefile:152: all] Error 2

--- a/src/views/view.h
+++ b/src/views/view.h
@@ -18,13 +18,13 @@
 
 #pragma once
 
+#include "common/history.h"
 #include "common/image.h"
 #ifdef HAVE_PRINT
 #include "common/cups_print.h"
 #endif
 #ifdef HAVE_MAP
 #include "common/geo.h"
-#include "common/history.h"
 #include <osm-gps-map.h>
 #endif
 #include <cairo.h>

